File: src/main/java/no/novari/integration/IntegrationService.java
package no.novari.integration;

import no.novari.integration.model.dtos.IntegrationDto;
import no.novari.integration.model.dtos.IntegrationPatchDto;
import no.novari.integration.model.dtos.IntegrationPostDto;
import no.novari.integration.model.entities.Integration;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Optional;
import java.util.Set;

@Service
public class IntegrationService {

    private final IntegrationRepository integrationRepository;
    private final IntegrationMappingService integrationMappingService;


    public IntegrationService(
            IntegrationRepository integrationRepository,
            IntegrationMappingService integrationMappingService
    ) {
        this.integrationRepository = integrationRepository;
        this.integrationMappingService = integrationMappingService;
    }

    public Collection<IntegrationDto> findAll() {
        return integrationMappingService.toDtos(
                integrationRepository.findAll()
        );
    }

    public Page<IntegrationDto> findAll(Pageable pageable) {
        return integrationRepository.findAll(pageable)
                .map(integrationMappingService::toDto);
    }

    public Collection<IntegrationDto> findAllBySourceApplicationIds(Set<Long> sourceApplicationIds) {
        return integrationMappingService.toDtos(
                integrationRepository.findIntegrationsBySourceApplicationIdIn(sourceApplicationIds)
        );
    }

    public Page<IntegrationDto> findAllBySourceApplicationIds(
            Set<Long> sourceApplicationIds,
            Pageable pageable
    ) {
        return integrationRepository.findIntegrationsBySourceApplicationIdIn(sourceApplicationIds, pageable)
                .map(integrationMappingService::toDto);
    }

    public boolean existsById(Long integrationId) {
        return integrationRepository.existsById(integrationId);
    }

    public Optional<IntegrationDto> findById(Long integrationId) {
        return integrationRepository.findById(integrationId)
                .map(integrationMappingService::toDto);
    }

    public Optional<IntegrationDto> findIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
            Long sourceApplicationId, String sourceApplicationIntegrationId
    ) {
        return integrationRepository.findIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
                        sourceApplicationId, sourceApplicationIntegrationId
                )
                .map(integrationMappingService::toDto);
    }

    public boolean existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
            Long sourceApplicationId, String sourceApplicationIntegrationId
    ) {
        return integrationRepository.existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
                sourceApplicationId, sourceApplicationIntegrationId
        );
    }

    public Optional<Long> findActiveConfigurationIdByIntegrationId(Long integrationId) {
        return integrationRepository.findById(integrationId)
                .map(Integration::getActiveConfigurationId);
    }

    public IntegrationDto save(IntegrationPostDto integrationPostDto) {
        return integrationMappingService.toDto(
                integrationRepository.save(
                        integrationMappingService.toIntegration(integrationPostDto)
                )
        );
    }

    public IntegrationDto updateById(Long integrationId, IntegrationPatchDto integrationPatchDto) {
        Integration integration = integrationRepository.getReferenceById(integrationId);

        integrationPatchDto.getDestination().ifPresent(integration::setDestination);
        integrationPatchDto.getState().ifPresent(integration::setState);
        integrationPatchDto.getActiveConfigurationId().ifPresent(integration::setActiveConfigurationId);

        return integrationMappingService.toDto(
                integrationRepository.save(integration)
        );
    }

}

File: src/main/kotlin/no/novari/flyt/integration/application/IntegrationService.kt
package no.novari.flyt.integration.application

import no.novari.flyt.integration.api.dto.IntegrationDto
import no.novari.flyt.integration.api.dto.IntegrationPatchDto
import no.novari.flyt.integration.api.dto.IntegrationPostDto
import no.novari.flyt.integration.persistence.IntegrationRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service

@Service
class IntegrationService(
    private val integrationRepository: IntegrationRepository,
    private val integrationMappingService: IntegrationMappingService,
) {
    fun findAll(): Collection<IntegrationDto> {
        return integrationMappingService.toDtos(integrationRepository.findAll())
    }

    fun findAll(pageable: Pageable): Page<IntegrationDto> {
        return integrationRepository.findAll(pageable).map(integrationMappingService::toDto)
    }

    fun findAllBySourceApplicationIds(sourceApplicationIds: Set<Long>): Collection<IntegrationDto> {
        return integrationMappingService.toDtos(
            integrationRepository.findIntegrationsBySourceApplicationIdIn(sourceApplicationIds),
        )
    }

    fun findAllBySourceApplicationIds(
        sourceApplicationIds: Set<Long>,
        pageable: Pageable,
    ): Page<IntegrationDto> {
        return integrationRepository
            .findIntegrationsBySourceApplicationIdIn(sourceApplicationIds, pageable)
            .map(integrationMappingService::toDto)
    }

    fun existsById(integrationId: Long): Boolean {
        return integrationRepository.existsById(integrationId)
    }

    fun findById(integrationId: Long): IntegrationDto? {
        return integrationRepository
            .findById(integrationId)
            .orElse(null)
            ?.let(integrationMappingService::toDto)
    }

    fun findIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
        sourceApplicationId: Long,
        sourceApplicationIntegrationId: String,
    ): IntegrationDto? {
        return integrationRepository
            .findIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
                sourceApplicationId,
                sourceApplicationIntegrationId,
            )?.let(integrationMappingService::toDto)
    }

    fun existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
        sourceApplicationId: Long,
        sourceApplicationIntegrationId: String,
    ): Boolean {
        return integrationRepository.existsIntegrationBySourceApplicationIdAndSourceApplicationIntegrationId(
            sourceApplicationId,
            sourceApplicationIntegrationId,
        )
    }

    fun findActiveConfigurationIdByIntegrationId(integrationId: Long): Long? {
        return integrationRepository.findById(integrationId).orElse(null)?.activeConfigurationId
    }

    fun save(integrationPostDto: IntegrationPostDto): IntegrationDto {
        return integrationMappingService.toDto(
            integrationRepository.save(integrationMappingService.toIntegration(integrationPostDto)),
        )
    }

    fun updateById(
        integrationId: Long,
        integrationPatchDto: IntegrationPatchDto,
    ): IntegrationDto {
        val integration = integrationRepository.getReferenceById(integrationId)

        integrationPatchDto.destination?.let { integration.destination = it }
        integrationPatchDto.state?.let { integration.state = it }
        integrationPatchDto.activeConfigurationId?.let { integration.activeConfigurationId = it }

        return integrationMappingService.toDto(integrationRepository.save(integration))
    }
}
